File: src/arrays/medium/move-number-to-end/c-sharp/in-place.cs
using System;
using System.Collections.Generic;

/*
Defines two pointers and if the current number is a target, just swaps it with the one at the end of the list, move the right counter, and analyze the left spot again. When the current number is not a target, move the left pointer to the right, until we cross the right pointer.

Time : O(n) - Where N is the length of the input list
Space: O(1) - Swap in place
*/
public class Program {
    public static List<int> MoveElementToEnd (List<int> array, int toMove) {
        int left = 0;
        int right = array.Count - 1;

        while (left < right) {
            if (array[left] == toMove) {
                array[left] = array[right];
                array[right] = toMove;
                right--;
            }
            else {
                left++;
            }
        }

        return array;
    }
}

File: src/arrays/medium/move-number-to-end/go/in-place.go
package main

// O(n) time | O(1) space - where N is the length of the input array
func MoveElementToEnd(array []int, toMove int) []int {
	i, j := 0, len(array)-1
	for i < j {
		for i < j && array[j] == toMove {
			j--
		}
		if array[i] == toMove {
			array[i], array[j] = array[j], array[i]
		}
		i++
	}
	return array
}
